octobot_trading/exchanges/parser/util.py

```diff
@@ -0,0 +1,424 @@
+        self.debugging_report_str = ""
+        self.missing_attributes_str = ""
+
+        self.raw_records = {}
+        self.exchange = exchange  # todo replace with logger
+        self.raw_record = {}
+        self.formatted_record = {}
+        self.formatted_records = []
+        self.multiple_records = False
+
+    def _ensure_details_completeness(self):
+        # ensure all REQUIRED_FIELDS are present and all NON_EMPTY_FIELDS are not empty
+
+        return all(
+            key in self.formatted_record for key in self.REQUIRED_FIELDS
+        ) and all(self.formatted_record[key] for key in self.NON_EMPTY_FIELDS)
+
+    def _create_debugging_report_for_record(self):
+        if not self.multiple_records:
+            self._create_debugging_report()
+
+    def _create_debugging_report(self):
+        if self.debugging_report_dict:
+            # filter duplicated errors
+            debugging_report_str = ""
+            missing_attributes_str = ""
+            for (
+                attribute_name,
+                error_messages_list,
+            ) in self.debugging_report_dict.items():
+                for index in range(len(error_messages_list)):
+                    if not error_messages_list[index] in error_messages_list[:index]:
+                        debugging_report_str += error_messages_list[index]
+                missing_attributes_str += f"{attribute_name}, "
+
+            self.exchange.logger.error(
+                self.debugging_report_template(
+                    f"Attributes with errors: {missing_attributes_str}\n"
+                    "\n"
+                    # f"Used exchange: {self.exchange}\n\n"  # todo add exchange name
+                    f"Debugging report: \n"
+                    "\n"
+                    f"{debugging_report_str}\n"
+                    f"The raw {self.PARSER_TITLE} response from the exchange: \n"
+                    "\n"
+                    f"{self.raw_records if self.multiple_records else self.raw_record}\n"
+                    "\n"
+                    f"Successfully parsed {self.PARSER_TITLE} attributes: \n"
+                    f"{self.formatted_records if self.multiple_records else self.formatted_record}\n"
+                )
+            )
+            raise NotImplementedError
+
+    def debugging_report_template(self, report) -> str:
+        return (
+            f"\n!!!!!!!!!!!!  DEBUGGING REPORT {self.PARSER_TITLE.upper()} PARSER START  !!!!!!!!!!!!\n"
+            "\n"
+            f"Post this report into OctoBot discord -> bug_report\n"
+            "^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n"
+            "This will help us to add support for your exchange\n"
+            "\n"
+            f"OctoBot is not able to parse all {self.PARSER_TITLE} attributes\n"
+            f"{report}"
+            "\n"
+            f"!!!!!!!!!!!!  DEBUGGING REPORT {self.PARSER_TITLE.upper()} PARSER END  !!!!!!!!!!!!"
+        )
+
+    def _try_to_find_and_set(
```

Comment:
I don't understand why we need so many "try to find" methods. CCXT and cryptofeed have documented standard keys, we can use them as default and when they're false or empty we do a specific implementation. I don't think it's a good idea to try to find by all means without really knowing what we used in the end.

Reply:
the main point of try to find is:
1. check validity of the value
2. log wrong or missing values, so that we get a useful debugging report
3. having a standardized way to set values that fit for all - so we have less to maintain

If there is just one key and everything is alright, it won't do much except checking and setting the value

try to find is a bit misleading, maybe "check_and_set_value" would be better

"I don't think it's a good idea to try to find by all means without really knowing what we used in the end."
We dont try by all means, just what is defined in each parser implementation. 
If there is anything wrong, we get a precise report on what we used, what failed and why it failed